source/MaterialXTest/CMakeLists.txt

```diff
@@ -1,41 +1,115 @@
-include_directories(
+set(MATERIALX_TEST_BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+
+# Discover all tests and allow them to be run in parallel (ctest -j20):
+function(add_tests _sources)
+  foreach(src_file ${_sources})
+    file(STRINGS ${src_file} matched_lines REGEX "TEST_CASE")
+    foreach(matched_line ${matched_lines})
+      string(REGEX REPLACE "(TEST_CASE[( \"]+)" "" test_name ${matched_line})
+      string(REGEX REPLACE "(\".*)" "" test_name ${test_name})
+      string(REGEX REPLACE "[^A-Za-z0-9_]+" "_" test_safe_name ${test_name})
+      add_test(NAME "MaterialXTest_${test_safe_name}"
+          COMMAND MaterialXTest ${test_name}
+          WORKING_DIRECTORY ${MATERIALX_TEST_BINARY_DIR})
+    endforeach()
+  endforeach()
+endfunction()
+
+add_executable(MaterialXTest)
+
+target_include_directories( MaterialXTest PUBLIC
     ${EXTERNAL_INCLUDE_DIRS}
     ${CMAKE_CURRENT_SOURCE_DIR}/../)
 
-file(GLOB_RECURSE materialx_source "${CMAKE_CURRENT_SOURCE_DIR}/*.cpp")
-file(GLOB_RECURSE materialx_headers "${CMAKE_CURRENT_SOURCE_DIR}/*.h")
+file(GLOB materialx_source "${CMAKE_CURRENT_SOURCE_DIR}/*.cpp")
+file(GLOB materialx_headers "${CMAKE_CURRENT_SOURCE_DIR}/*.h")
+
+list(APPEND materialx_headers "${CMAKE_CURRENT_SOURCE_DIR}/Catch/catch.hpp")
 
 assign_source_group("Source Files" ${materialx_source})
 assign_source_group("Header Files" ${materialx_headers})
 
-add_executable(MaterialXTest ${materialx_source} ${materialx_headers})
+target_sources(MaterialXTest PUBLIC ${materialx_source} ${materialx_headers})
 
-# Discover all tests and allow them to be run in parallel (ctest -j20):
-foreach(src_file ${materialx_source})
-    file(STRINGS ${src_file} matched_lines REGEX "TEST_CASE")
-    foreach(matched_line ${matched_lines})
-        string(REGEX REPLACE "(TEST_CASE[( \"]+)" "" test_name ${matched_line})
-        string(REGEX REPLACE "(\".*)" "" test_name ${test_name})
-        string(REGEX REPLACE "[^A-Za-z0-9_]+" "_" test_safe_name ${test_name})
-        add_test(NAME "MaterialXTest_${test_safe_name}"
-            COMMAND MaterialXTest ${test_name}
-            WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
-    endforeach()
-endforeach()
+add_subdirectory(MaterialXCore)
+target_link_libraries(MaterialXTest MaterialXCore)
+add_subdirectory(MaterialXFormat)
+target_link_libraries(MaterialXTest MaterialXFormat)
+
+if(MATERIALX_BUILD_GEN)
+  add_subdirectory(MaterialXGenShader)
+  target_link_libraries(MaterialXTest MaterialXGenShader)
+  if(MATERIALX_BUILD_RUNTIME)
+    add_subdirectory(MaterialXRuntime)
+    target_link_libraries(MaterialXTest MaterialXRuntime)
```

Comment:
Just double checking here. We don't need a cmake version bump for this, or any other calls?

Reply:
No, this should work in 3.1. Multiple calls are allowed, and the linked library target need not be declared in same directory. It definitely works on the build slaves.